src/machine/machine_stm32f103xx.go

```diff
@@ -290,3 +291,463 @@ func (spi SPI) setPins(sck, mosi, miso uint8) {
+			}
+
+			// clear address here
+			timeout := TIMEOUT
+			for i2c.Bus.SR2&(stm32.I2C_SR2_MSL|stm32.I2C_SR2_BUSY) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read clear address")
+				}
+			}
+
+			for i := 0; i < len(r)-3; i++ {
+				// Enable ACK of received data
+				i2c.Bus.CR1 |= stm32.I2C_CR1_ACK
+
+				// wait for btf. we need a longer timeout here than normal.
+				timeout = 1000
+				for (i2c.Bus.SR1 & stm32.I2C_SR1_BTF) == 0 {
+					timeout--
+					if timeout == 0 {
+						println("I2C timeout on read 3 bytes")
+						return errors.New("I2C timeout on read 3 bytes")
+					}
+				}
+
+				// read the next byte
+				r[i] = byte(i2c.Bus.DR)
+			}
+
+			// wait for btf. we need a longer timeout here than normal.
+			timeout = 1000
+			for (i2c.Bus.SR1 & stm32.I2C_SR1_BTF) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read more than 3 bytes")
+				}
+			}
+
+			// Disable ACK of received data
+			i2c.Bus.CR1 &^= stm32.I2C_CR1_ACK
+
+			// get third from last byte
+			r[len(r)-3] = byte(i2c.Bus.DR)
+
+			// Generate stop condition
+			i2c.Bus.CR1 |= stm32.I2C_CR1_STOP
+
+			// get second from last byte
+			r[len(r)-2] = byte(i2c.Bus.DR)
+
+			timeout = TIMEOUT
+			for (i2c.Bus.SR1 & stm32.I2C_SR1_RxNE) == 0 {
+				timeout--
+				if timeout == 0 {
+					return errors.New("I2C timeout on read last byte of more than 3")
+				}
+			}
+
+			// get last byte
+			r[len(r)-1] = byte(i2c.Bus.DR)
+
+			// wait for stop
+			return i2c.waitForStop()
+		}
+	}
+
+	return nil
+}
+
+const TIMEOUT = 500
+
+// signalStart sends a start signal.
+func (i2c I2C) signalStart() error {
+	// Wait until I2C is not busy
+	timeout := TIMEOUT
+	for (i2c.Bus.SR2 & stm32.I2C_SR2_BUSY) > 0 {
+		timeout--
+		if timeout == 0 {
+			return errors.New("I2C busy on start")
+		}
+	}
+
+	// clear stop
+	i2c.Bus.CR1 &^= stm32.I2C_CR1_STOP
+
+	// Generate start condition
+	i2c.Bus.CR1 |= stm32.I2C_CR1_START
+
+	// Wait for I2C EV5 aka SB flag.
+	timeout = TIMEOUT
+	for (i2c.Bus.SR1 & stm32.I2C_SR1_SB) == 0 {
+		timeout--
+		if timeout == 0 {
+			return errors.New("I2C timeout on start")
+		}
+	}
+
+	return nil
+}
+
+// signalStop sends a stop signal and waits for it to succeed.
+func (i2c I2C) signalStop() error {
+	// Generate stop condition
+	i2c.Bus.CR1 |= stm32.I2C_CR1_STOP
+
+	// wait for stop
+	return i2c.waitForStop()
+}
+
+// waitForStop waits after a stop signal.
+func (i2c I2C) waitForStop() error {
+	// Wait until I2C is stopped
+	timeout := TIMEOUT
+	for (i2c.Bus.SR1 & stm32.I2C_SR1_STOPF) > 0 {
+		timeout--
+		if timeout == 0 {
+			println("I2C timeout on wait for stop signal")
+			return errors.New("I2C timeout on wait for stop signal")
+		}
+	}
+
+	return nil
+}
+
+// Send address of device we want to talk to
+func (i2c I2C) sendAddress(address uint8, write bool) error {
+	data := (address << 1)
+	if write {
+		data &^= stm32.I2C_OAR1_ADD0
+	} else {
+		data |= stm32.I2C_OAR1_ADD0
+	}
```

Comment:
I _guess_ this is here to set the read/write bit? If so, it is probably equivalent to the code [here](https://github.com/aykevl/tinygo/blob/master/src/machine/machine_atmega.go#L174).
Totally optional, but I think the code there is more readable.

Reply:
good call, I changed to match the more readable code.